### ISSUE TITLE
Error in Algorithm 1 of Flash Attention 2 paper

### ISSUE THREAD

**Issue by mbchang** [Dao-AILab/flash-attention#991]
On line 10 of Algorithm 1, FlashAttention-2 forward pass, it says
![image](https://github.com/Dao-AILab/flash-attention/assets/6439365/0c646c47-6f1e-4a60-a943-a15b58c825ff)

However, $\text{diag}\left(e^{m_i^{j-1} - m_i^{j}}\right)^{-1}\mathbf{O}_i^{(j-1)}$ should not have the $^{-1}$ and actually just be $\text{diag}\left(e^{m_i^{j-1} - m_i^{j}}\right)\mathbf{O}_i^{(j-1)}$.

Similar the online softmax trick example at the top of page 6 should also not have the $^{-1}$ before the $\tilde{\mathbf{O}}^{(1)}$
![image](https://github.com/Dao-AILab/flash-attention/assets/6439365/b330a1cc-d622-4f0a-9eb2-7802c7aeb620)

Lastly, there is a typo, where the following line
![image](https://github.com/Dao-AILab/flash-attention/assets/6439365/6e36f98e-596d-415b-b2c7-54291482a836)
should be deleted from the online softmax trick derivation in page 6. It looks like it was copied from the online softmax trick derivation on page 4 for Flash Attention 1 and was not removed.

**Comment by jxjalsq** [comment]
when i impl flash attention  v2 in android with c++, spilt k, v, q with block size is 1, the last result is pricesion with 10e-2, the error is so big.

**Comment by xyg-coder** [comment]
I was looking at triton [implementation](https://triton-lang.org/main/getting-started/tutorials/06-fused-attention.html) and it matches your idea. Thanks for pointing this out.

**Comment by AtomicVar** [comment]
I agree with you. In addition, I also think there's a typo on Page 4 (describing FlashAttention 1):

<img width="739" alt="image" src="https://github.com/user-attachments/assets/30114b63-4425-46ce-836c-9718170f1b4e">

The $^{-1}$ should be deleted and it missed $e^{m^{(1)}-m^{(2)}}$ before $O^{(1)}$

**Comment by ahban** [comment]
Why did the author delete his response? Is there any information we missed? :(

**Comment by Victor-Jung** [comment]
Oh man thanks a lot for spotting this issue! 😁 I was reproducing the algorithm and couldn't understand why I wasn't getting the expected output! I think it's pretty important to get this typo fixed in the arxiv paper read by so many (@tridao).

**Comment by Victor-Jung** [comment]
This typo is also present in the more recent [Flash Attention 3](https://arxiv.org/pdf/2407.08608) paper. In Algorithm 1, line 21.

<img width="516" height="29" alt="Image" src="https://github.com/user-attachments/assets/9d766033-a5a5-4752-bd72-c2151bda09b7" />

**Comment by parikshit14** [comment]
I spent two days figuring out why $^{-1}$ was there..... and finally i found this post
Thanks for posting.
